### PR TITLE
Support configurable warnings for planner=COST with Updates

### PR DESCRIPTION
**Pull request by craigtaverner**

### CODE REVIEW

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/notification/InternalNotification.scala [hunk 1]
```diff
@@ -31,3 +31,5 @@ case class CartesianProductNotification(position: InputPosition) extends Interna
 case class LengthOnNonPathNotification(position: InputPosition) extends InternalNotification
 
 case object LegacyPlannerNotification extends InternalNotification
+
+case object CostForUpdatesNotification extends InternalNotification
```
Comment (davidegrohmann) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/notification/InternalNotification.scala]: I would suggest something like `HintedPlannerNotApplicableNotification` instead. Since this kind of warning is unrelated to `updates` (it just happens that at the moment we do not support update queries in cost).

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/CypherCompiler.scala [hunk 2]
```diff
@@ -83,9 +84,14 @@ object CypherCompilerFactory {
...
-    val pipeBuilder = pickedPlannerName match {
-      case FallbackPlannerName =>
+    val pipeBuilder = (pickedPlannerName, useErrorsOverWarnings) match {
+      // fallback to RULE if specific planner was not specified
+      case (FallbackPlannerName, _) =>
         new LegacyVsNewExecutablePlanBuilder(new LegacyExecutablePlanBuilder(monitors, rewriterSequencer), planner, planBuilderMonitor)
+      // fallback to RULE if configuration does not require error if specified planner cannot be supported
+      case (_, false) =>
+        new LegacyVsNewExecutablePlanBuilder(new LegacyExecutablePlanBuilder(monitors, rewriterSequencer), planner, planBuilderMonitor)
+      // generate error if specified planner cannot be supported
```
Comment (davidegrohmann) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/CypherCompiler.scala]: perhaps it is easier to use an `if-then-else` here:

```
if (pickedPlannerName == FallbackPlannerName || !useErrorsOverWarnings)
  new LegacyVsNewExecutablePlanBuilder(...)
else
  new ErrorReportingExecutablePlanBuilder(...)
```

File: community/cypher/cypher/src/test/scala/org/neo4j/cypher/CypherCompatibilityTest.scala [hunk 3]
```diff
@@ -20,6 +20,9 @@
 package org.neo4j.cypher
 
 import java.util
+import org.neo4j.cypher.internal.RewindableExecutionResult
+import org.neo4j.cypher.internal.compatibility.ExecutionResultWrapperFor2_3
+import org.neo4j.cypher.internal.compiler.v2_3.notification.{InternalNotification, CostForUpdatesNotification}
```
Comment (davidegrohmann) [inline community/cypher/cypher/src/test/scala/org/neo4j/cypher/CypherCompatibilityTest.scala]: I think in this test suite you should not use the compiler internal notifications

File: community/kernel/src/main/java/org/neo4j/graphdb/factory/GraphDatabaseSettings.java [hunk 4]
```diff
@@ -94,6 +94,9 @@
             "dbms.cypher.planner",
             options( "COST", "RULE", DEFAULT ), DEFAULT );
 
+    @Description( "Set this to specify the behavior for conformance to planner hints." )
```
Comment (davidegrohmann) [inline community/kernel/src/main/java/org/neo4j/graphdb/factory/GraphDatabaseSettings.java]: I would remove the `planner` from description, since the property name doesn't suggest coupling with planning.

File: community/kernel/src/main/java/org/neo4j/graphdb/impl/notification/NotificationCode.java [hunk 5]
```diff
@@ -40,6 +40,9 @@
     LEGACY_PLANNER( SeverityLevel.WARNING,
                     Status.Statement.DeprecationWarning,
                     "Using PLANNER for switching between planners has been deprecated, please use CYPHER planner=[rule,cost] instead"),
+    COST_FOR_UPDATES( SeverityLevel.WARNING,
+                    Status.Statement.CostForUpdatesWarning,
+                    "Using COST planner for update queries is not supported, please use RULE planner instead"),
```
Comment (davidegrohmann) [inline community/kernel/src/main/java/org/neo4j/graphdb/impl/notification/NotificationCode.java]: Same as `CostForUpdatesNotification`.

File: community/kernel/src/main/java/org/neo4j/kernel/api/exceptions/Status.java [hunk 6]
```diff
@@ -159,6 +159,7 @@ private Transaction( Classification classification, String description )
         ExecutionFailure( DatabaseError, "The database was unable to execute the statement." ),
         ExternalResourceFailure( TransientError, "The external resource is not available"),
         CartesianProduct( ClientNotification, "This query builds a cartesian product between disconnected patterns." ),
+        CostForUpdatesWarning( ClientNotification, "This query requests the COST planner for updates." ),
```
Comment (davidegrohmann) [inline community/kernel/src/main/java/org/neo4j/kernel/api/exceptions/Status.java]: Same as `CostForUpdatesNotification`.

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/CypherCompiler.scala [hunk 7]
```diff
@@ -83,9 +84,14 @@ object CypherCompilerFactory {
...
     // falling back to legacy planner is allowed only when no cost-based planner is picked explicitly (e.g., COST, IDP)
-    val pipeBuilder = pickedPlannerName match {
-      case FallbackPlannerName =>
+    val pipeBuilder = (pickedPlannerName, useErrorsOverWarnings) match {
+      // fallback to RULE if specific planner was not specified
+      case (FallbackPlannerName, _) =>
         new LegacyVsNewExecutablePlanBuilder(new LegacyExecutablePlanBuilder(monitors, rewriterSequencer), planner, planBuilderMonitor)
+      // fallback to RULE if configuration does not require error if specified planner cannot be supported
+      case (_, false) =>
+        new LegacyVsNewExecutablePlanBuilder(new LegacyExecutablePlanBuilder(monitors, rewriterSequencer), planner, planBuilderMonitor)
```
Comment (davidegrohmann) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/CypherCompiler.scala]: Here you need `WarningReportingExecutablePlanBuilder` similar to the `ErrorReportingExecutablePlanBuilder`

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/notification/InternalNotification.scala (hunk 1 above)
Comment (craigtaverner) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/notification/InternalNotification.scala]: The thinking at the time was that this was a temporary feature that would 'soon' be removed, and so making it specific to the current case made sense. For the runtime we went the generic route, because we have no idea how long the transition from interpreted->compiled will take and the border will constantly move.

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/CypherCompiler.scala (hunk 2 above)
Comment (craigtaverner) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/CypherCompiler.scala]: Based on our discussion early today we've re-written this completely to match the more correct logic (including not showing warnings when using defaults).

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/CypherCompiler.scala (hunk 7 above)
Comment (craigtaverner) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/CypherCompiler.scala]: The behaviour (and code) of the suggested `WarningReportingExecutablePlanBuilder` is almost identical to the `LegacyVsNewExecutablePlanBuilder`, with the single addition of a warning, so we choose to simply pass a boolean into the latter to warn, or be silent.

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/notification/InternalNotification.scala (hunk 1 above)
Comment (boggle) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/notification/InternalNotification.scala]: CostForUpdates seems confusing to me, had no clue what it would be about when reading the name.

File: community/kernel/src/main/java/org/neo4j/graphdb/factory/GraphDatabaseSettings.java [hunk 8]
```diff
@@ -94,6 +94,9 @@
             "dbms.cypher.planner",
             options( "COST", "RULE", DEFAULT ), DEFAULT );
 
+    @Description( "Set this to specify the behavior for conformance to Cypher hints." )
```
Comment (boggle) [inline community/kernel/src/main/java/org/neo4j/graphdb/factory/GraphDatabaseSettings.java]: Perhaps this could be worded with a bit more detail? These strings show up in the manual and thus should be explicit and clear.

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/planner/CostBasedExecutablePlanBuilder.scala [hunk 9]
```diff
@@ -69,35 +72,36 @@ case class CostBasedExecutablePlanBuilder(monitors: Monitors,
         val (logicalPlan, pipeBuildContext) = produceLogicalPlan(ast, rewrittenSemanticTable)(planContext)
         monitor.foundPlan(inputQuery.queryText, logicalPlan)
         try {
-          val res = produceCompiledPlan(logicalPlan, inputQuery, rewrittenSemanticTable,
-            planContext, pipeBuildContext, planBuilderMonitor)
+          val res = runtimeName match {
+            case CompiledRuntimeName => produceCompiledPlan(logicalPlan, rewrittenSemanticTable, planContext, planBuilderMonitor)
+            case _ => produceInterpretedPlan(logicalPlan, pipeBuildContext, planContext)
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/planner/CostBasedExecutablePlanBuilder.scala]: Does this mean we go back to having interpreted as default runtime again?

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/planner/CostBasedExecutablePlanBuilder.scala [hunk 10]
```diff
@@ -69,35 +72,36 @@ case class CostBasedExecutablePlanBuilder(monitors: Monitors,
...
-        val codeGen = new CodeGenerator
-        Left(codeGen.generate(logicalPlan, planContext, clock, semanticTable))
-    }
+  private def produceInterpretedPlan(logicalPlan: LogicalPlan, pipeBuildContext: PipeExecutionBuilderContext, planContext: PlanContext) =
+    Right(executionPlanBuilder.build(logicalPlan)(pipeBuildContext, planContext))
+
+  private def produceCompiledPlan(logicalPlan: LogicalPlan, semanticTable: SemanticTable, planContext: PlanContext, monitor: NewRuntimeSuccessRateMonitor) = {
+    monitor.newPlanSeen(logicalPlan)
+    val codeGen = new CodeGenerator
+    Left(codeGen.generate(logicalPlan, planContext, clock, semanticTable))
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/planner/CostBasedExecutablePlanBuilder.scala]: We should be able to inline `codeGen` here.

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/RuntimeName.scala [hunk 11]
```diff
@@ -32,6 +32,10 @@ case object CompiledRuntimeName extends RuntimeName {
   override val name = "COMPILED"
 }
 
+case object DefaultRuntimeName extends RuntimeName {
+  override val name = "DEFAULT"
+}
+
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/RuntimeName.scala:38]: Why is the `default` val on line 41 still the `InterpretedRuntimeName` if we now have a `DefaultRuntimeName`?

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/planner/CostBasedPipeBuilderFactory.scala [hunk 12]
```diff
@@ -38,7 +38,9 @@ object CostBasedPipeBuilderFactory {
             tokenResolver: SimpleTokenResolver = new SimpleTokenResolver(),
             maybeExecutionPlanBuilder: Option[PipeExecutionPlanBuilder] = None,
             plannerName: CostBasedPlannerName = PlannerName.default,
-            runtimeName: RuntimeName = RuntimeName.default
+            runtimeName: RuntimeName = RuntimeName.default,
+            useErrorsOverWarnings: Boolean = false,
+            showWarnings: Boolean = false
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/planner/CostBasedPipeBuilderFactory.scala:42]: I don't really like adding boolean parameters, especially not after each other. Couldn't we send through a function `handleFallback` which will take some enum parameter `severity`, with modes `error`, `warning` and `silence`, and produces an `Option` instead? It would also take the function that produces the fallback plan in case `error` wasn't used.

File: community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/CypherCompiler.scala [hunk 13]
```diff
@@ -67,6 +67,7 @@ class CypherCompiler(graph: GraphDatabaseService,
                      configuredVersion: CypherVersion,
                      configuredPlanner: CypherPlanner,
                      configuredRuntime: CypherRuntime,
+                     useErrorsOverWarnings: Boolean,
```
Comment (Mats-SX) [inline community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/CypherCompiler.scala:68]: A similar situation applies to this parameter too, as above.

File: community/kernel/src/main/java/org/neo4j/graphdb/impl/notification/NotificationCode.java [hunk 14]
```diff
@@ -40,6 +40,12 @@
     LEGACY_PLANNER( SeverityLevel.WARNING,
                     Status.Statement.DeprecationWarning,
                     "Using PLANNER for switching between planners has been deprecated, please use CYPHER planner=[rule,cost] instead"),
+    PLANNER_UNSUPPORTED( SeverityLevel.WARNING,
+                    Status.Statement.PlannerUnsupportedWarning,
+                    "Using COST planner is unsupported for this query, please use RULE planner instead"),
+    RUNTIME_UNSUPPORTED( SeverityLevel.WARNING,
+                    Status.Statement.RuntimeUnsupportedWarning,
+                    "Using compiled runtime is unsupported for this query, please use interpreted runtime instead"),
```
Comment (Mats-SX) [inline community/kernel/src/main/java/org/neo4j/graphdb/impl/notification/NotificationCode.java]: Is the recommended syntax capitals for planners and minuscules for runtimes? Otherwise, we should be consistent here and print `RUNTIME` and `INTERPRETED` instead.

**Comment by systay** [comment]
Also, we need to rebase this bad boy

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/planner/CostBasedExecutablePlanBuilder.scala (hunk 9 above)
Comment (davidegrohmann) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/planner/CostBasedExecutablePlanBuilder.scala]: This logic is completely broken, we should never fallback if you explicitly ask for `CompiledRuntimeName`, we need to fix the behavior here (which is btw not related to this PR).

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/planner/CostBasedPipeBuilderFactory.scala (hunk 12 above)
Comment (davidegrohmann) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/planner/CostBasedPipeBuilderFactory.scala:42]: I am +1 on this, but let's make this biggish change in a separate PR, I don't like to defer the merge further.

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/planner/CostBasedExecutablePlanBuilder.scala (hunk 9 above)
Comment (pontusmelke) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/planner/CostBasedExecutablePlanBuilder.scala]: We should fallback if `useErrorsOverWarnings == false` right? Not sure I agree that it is not part of the PR, the code has been changed from always falling back to sometimes falling back and in the process it looks like we changed the default to being interpreted again?

**Comment by craigtaverner** [comment]
Rebased to latest 2.3, hopefully a merge can work now
